jmclient/jmclient/taker.py

```diff
@@ -990,3 +956,40 @@ def round_to_significant_figures(d, sf):
             sigfiged = int(round(d/power10*sf_power10)*power10/sf_power10)
             return sigfiged
     raise RuntimeError()
+
+def calculate_fidelity_bond_values(fidelity_bonds_info):
+    interest_rate = get_interest_rate()
+    blocks = jm_single().bc_interface.get_current_block_height()
+    mediantime = jm_single().bc_interface.get_best_block_median_time()
+
+    parsed_bonds = [FidelityBondProof.parse_and_verify_proof_msg(fb["counterparty"],
+        fb["takernick"], fb["proof"]) for fb in fidelity_bonds_info]
+    validated_bonds = [FidelityBondMixin.get_validated_timelocked_fidelity_bond_utxo(
+        pb.utxo, pb.utxo_pub, pb.locktime, pb.cert_expiry, blocks)
+        for pb in parsed_bonds if pb != None]
+
+    validated_fidelity_bond_data_with_dups = [(bond_data, utxo_data)
+        for bond_data, utxo_data in zip(parsed_bonds, validated_bonds)
+        if utxo_data != None]
+    #check for duplicated utxos i.e. two or more makers using the same UTXO
+    # which is obviously not allowed, a fidelity bond must only be usable by one maker nick
```

Comment:
If I understand correctly, the following code just drops/ignores duplicates. Is there a case for banning both/all duplicates in case we have valid FB proofs coming in on more than one maker nick?

Reply:
I don't think so. We can't ban just one because we can't know which fidelity bond message arrived first. Dropping duplicates more-or-less randomly removes the incentive to announce a fidelity bond on more than one maker, which should be enough.

Reply:
> We can't ban just one because we can't know which fidelity bond message arrived first.

Yes but I was suggesting banning *both*, not one, on the basis that since they're both validated, they definitely represent one entity trying to reuse.

Reply:
Yes I see.

I think it doesn't make much difference either way. I suppose I thought a maker might accidentality announce twice and so it would be good to err on the side of being lax. If the fidelity bond UTXO is accepted once then the joinmarket ecosystem still benefits from it because it still makes sybil attackers more costly.